### ISSUE TITLE
CBLK: NVMe action returns same slot/req multiple times

### ISSUE THREAD

**Issue by fhaverkamp** [open-power/snap#470]
commit 20785837f0a241c8bd9c620c219ba8c9e5562640 (HEAD -> action_hdl_nvme_example, origin/action_hdl_nvme_example)
Author: Frank Haverkamp <haver@linux.vnet.ibm.com>
Date:   Mon Oct 2 15:31:48 2017 +0200

    CBLK: Adding timeouts and other stuff to see if parallelism works OK or not
    
    Signed-off-by: Frank Haverkamp <haver@linux.vnet.ibm.com>

Single threaded use works ok so far:
haver@tul3eth3:~/snap/software$ ./tools/snap_init.sh -C0 -TCBLK

I tried now to read in parallel 5 threads 20 blocks and see:

```
$ make all && SNAP_TRACE=0x20 LD_LIBRARY_PATH=lib ./tools/snap_cblk -C0 -b1 -s0 -n20 -t5 --read cblk_read2.bin 2> bla.log

haver@tul3eth3:~/snap/software$ egrep -e HW -e err bla.log 
B 0000876c.0000876e 1506950386030        [start_memcpy] HW COPY_NH memcpy_104(slot=1 dest=0x3fffb1e30000, src=0x0 n=4096 bytes)
B 0000876c.0000876d 1506950386273        [read_status] HW READ_COMPLETED 00000011 slot: 1
B 0000876c.00008770 1506950386437        [start_memcpy] HW COPY_NH memcpy_104(slot=1 dest=0x3fffb1e30000, src=0x8 n=4096 bytes)
B 0000876c.00008772 1506950386483        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3fffb1e50000, src=0x18 n=4096 bytes)
B 0000876c.0000876d 1506950386529        [read_status] HW READ_COMPLETED 00000011 slot: 1
B 0000876c.0000876f 1506950386605        [start_memcpy] HW COPY_NH memcpy_304(slot=3 dest=0x3fffb1e70000, src=0x10 n=4096 bytes)
B 0000876c.0000876d 1506950386614        [read_status] HW READ_COMPLETED 00000012 slot: 2
B 0000876c.0000876e 1506950386778        [start_memcpy] HW COPY_NH memcpy_404(slot=4 dest=0x3fffb1e90000, src=0x28 n=4096 bytes)
B 0000876c.0000876d 1506950386742        [read_status] HW READ_COMPLETED 00000013 slot: 3
B 0000876c.00008771 1506950386952        [start_memcpy] HW COPY_NH memcpy_504(slot=5 dest=0x3fffb1eb0000, src=0x20 n=4096 bytes)
B 0000876c.00008770 1506950386986        [start_memcpy] HW COPY_NH memcpy_604(slot=6 dest=0x3fffb1ed0000, src=0x30 n=4096 bytes)
B 0000876c.0000876d 1506950387078        [read_status] HW READ_COMPLETED 00000014 slot: 4
B 0000876c.00008772 1506950387143        [start_memcpy] HW COPY_NH memcpy_704(slot=7 dest=0x3fffb1ef0000, src=0x38 n=4096 bytes)
B 0000876c.0000876d 1506950387259        [read_status] HW READ_COMPLETED 00000015 slot: 5
B 0000876c.0000876d 1506950387299        [read_status] HW READ_COMPLETED 00000016 slot: 6
B 0000876c.0000876f 1506950387411        [start_memcpy] HW COPY_NH memcpy_804(slot=8 dest=0x3fffb1f10000, src=0x40 n=4096 bytes)
B 0000876c.0000876d 1506950387373        [read_status] HW READ_COMPLETED 00000017 slot: 7
B 0000876c.0000876e 1506950387503        [start_memcpy] HW COPY_NH memcpy_904(slot=9 dest=0x3fffb1f30000, src=0x48 n=4096 bytes)
B 0000876c.00008771 1506950387690        [start_memcpy] HW COPY_NH memcpy_a04(slot=10 dest=0x3fffb1f50000, src=0x50 n=4096 bytes)
B 0000876c.00008770 1506950387758        [start_memcpy] HW COPY_NH memcpy_b04(slot=11 dest=0x3fffb1f70000, src=0x58 n=4096 bytes)
B 0000876c.0000876d 1506950387659        [read_status] HW READ_COMPLETED 00000018 slot: 8
B 0000876c.00008772 1506950387924        [start_memcpy] HW COPY_NH memcpy_c04(slot=12 dest=0x3fffb1f90000, src=0x60 n=4096 bytes)
B 0000876c.0000876d 1506950387875        [read_status] HW READ_COMPLETED 00000019 slot: 9
B 0000876c.0000876d 1506950387983        [read_status] HW READ_COMPLETED 0000001a slot: 10
B 0000876c.0000876d 1506950388052        [read_status] HW READ_COMPLETED 0000001b slot: 11
B 0000876c.0000876d 1506950388148        [read_status] HW READ_COMPLETED 0000001c slot: 12
B 0000876c.0000876f 1506950388452        [start_memcpy] HW COPY_NH memcpy_d04(slot=13 dest=0x3fffb1fb0000, src=0x68 n=4096 bytes)
B 0000876c.0000876e 1506950388501        [start_memcpy] HW COPY_NH memcpy_e04(slot=14 dest=0x3fffb1fd0000, src=0x70 n=4096 bytes)
B 0000876c.00008771 1506950388551        [start_memcpy] HW COPY_NH memcpy_f04(slot=15 dest=0x3fffb1ff0000, src=0x78 n=4096 bytes)
B 0000876c.00008772 1506950388677        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3fffb1e50000, src=0x88 n=4096 bytes)
B 0000876c.00008770 1506950388626        [start_memcpy] HW COPY_NH memcpy_104(slot=1 dest=0x3fffb1e30000, src=0x80 n=4096 bytes)
B 0000876c.0000876d 1506950388752        [read_status] HW READ_COMPLETED 0000001f slot: 15
B 0000876c.0000876d 1506950388812        [read_status] HW READ_COMPLETED 00000012 slot: 2
B 0000876c.0000876d 1506950388855        [read_status] HW READ_COMPLETED 0000001f slot: 15
B 0000876c.0000876d 1506950388880      [done_thread] err: slot 15 status is READY, in_flight=0/4 ILLEGAL STATUS (1165)
B 0000876c.0000876d 1506950388896        [read_status] HW READ_COMPLETED 00000012 slot: 2
B 0000876c.0000876d 1506950388962      [done_thread] err: slot 2 status is READY, in_flight=0/3 ILLEGAL STATUS (1165)
B 0000876c.0000876d 1506950388980        [read_status] HW READ_COMPLETED 00000011 slot: 1
B 0000876c.00008771 1506950389143        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3fffb1e50000, src=0x90 n=4096 bytes)
B 0000876c.00008772 1506950389242        [start_memcpy] HW COPY_NH memcpy_304(slot=3 dest=0x3fffb1e70000, src=0x98 n=4096 bytes)
B 0000876c.0000876d 1506950389327        [read_status] HW READ_COMPLETED 00000012 slot: 2
B 0000876c.0000876d 1506950389363        [read_status] HW READ_COMPLETED 00000013 slot: 3
B 0000876c.0000876d 1506950394001      err: req[13]: READING 0 5000/5001 TIMEOUT!
B 0000876c.0000876d 1506950394014      err: req[14]: READING 0 5000/5001 TIMEOUT!
err: cblk_read unhappy rc=0! Timer expired
err: cblk_read unhappy rc=0! Timer expired
haver@tul3eth3:~/snap/software$ 

```

**Comment by fhaverkamp** [comment]
What is shown here is that I think that I am missing completions for slots/reqs 13 and 14.
Instead I see double completions for 15 and 2. That confuses the software at this point in time, I do not repeat failing requests. I suspect that the NVMe action responsible for the queuing gets out of sync somehow. Let talk on Wednesday to see what is going on there.

**Comment by fhaverkamp** [comment]
Thanks Sven for having a look at this. The new hardware version appears working correctly. I needed to add some more software changes which fix some remaining issues.

The following branch should have both correct: software as well as hardware:
    https://github.com/open-power/snap/tree/hdl_nvme_action

Let us consider merging this.

**Comment by fhaverkamp** [comment]
I am running the following test-case successfully:
```
haver@tul3eth3:~/snap/software$ for t in 1 8 16 32 ; do ./tools/snap_init.sh -C0 -t$t -TCBLK ; done
```
It prints out some performance values too.
```
SNAP on FGT Card, NVME enabled, 4096 MB SRAM available.
SNAP FPGA Release: v1.1.1 Distance: 44 GIT: 0xb8d81fd7
SNAP FPGA Build (Y/M/D): 2017/10/05 Time (H:M): 17:41
SNAP FPGA CIR Master: 1 My ID: 0
SNAP FPGA Up Time: 25630 sec
SNAP FPGA Exploration already done (MSAT: 1 MAID: 1)

   Short |  Action Type |   Level   |
   ------+--------------+-----------+-----------
     0     0x10140001     0x00000000  IBM HDL NVMe example (only for FGT card)

Init SSD[0]
     SSD[0] InitDone: 1 QueueDone: 1
     SSD[1] InitDone: 1 QueueDone: 1
SSD[0] PCIE  -> UP (after 0 sec).
SSD[0] PCIE Link Rate: Gen3 Link With: 4
     SSD[0] Capability Register: 0x0
Init SSD[1]
     SSD[0] InitDone: 1 QueueDone: 1
     SSD[1] InitDone: 1 QueueDone: 1
SSD[1] PCIE  -> UP (after 0 sec).
SSD[1] PCIE Link Rate: Gen3 Link With: 4
     SSD[1] Capability Register: 0x0
Exit rc: 0
### (1.1) Formatting using 1 block increasing pattern ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Formatting NVMe drive 1048576 KiB with pattern 01 ...
Formatting of 1073741824 bytes took 6453293 usec @ 166.387 MiB/sec
# Reading using 32 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 8192 times 131072 bytes: 1048576 KiB NVMe into cblk_read.bin
Reading of 1073741824 bytes took 956824 usec @ 1122.194 MiB/sec
NOTE: Please manually inspect if pattern is really 1
0000000 0101 0101 0101 0101 0101 0101 0101 0101
*
40000000

### (1.2) Formatting using 2 block increasing pattern ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Formatting NVMe drive 1048576 KiB with pattern 02 ...
Formatting of 1073741824 bytes took 4273376 usec @ 251.263 MiB/sec
# Reading using 32 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 8192 times 131072 bytes: 1048576 KiB NVMe into cblk_read.bin
Reading of 1073741824 bytes took 957603 usec @ 1121.281 MiB/sec
NOTE: Please manually inspect if pattern is really 2
0000000 0202 0202 0202 0202 0202 0202 0202 0202
*
40000000

### (2) Formatting using 2 blocks increasing pattern ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Formatting NVMe drive 1048576 KiB with pattern ff ...
Formatting of 1073741824 bytes took 4269098 usec @ 251.515 MiB/sec
# Reading using 1 block ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 262144 times 4096 bytes: 1048576 KiB NVMe into cblk_read1.bin
Reading of 1073741824 bytes took 1919037 usec @ 559.521 MiB/sec
# Reading using 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 131072 times 8192 bytes: 1048576 KiB NVMe into cblk_read2.bin
Reading of 1073741824 bytes took 1113125 usec @ 964.619 MiB/sec
Compare results ...

### (3.1) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4268462 usec @ 251.552 MiB/sec
# Reading 1 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 262144 times 4096 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 1906674 usec @ 563.149 MiB/sec
Compare results ...

### (3.2) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4272434 usec @ 251.319 MiB/sec
# Reading 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 131072 times 8192 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 1167998 usec @ 919.301 MiB/sec
Compare results ...

### (3.4) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4316793 usec @ 248.736 MiB/sec
# Reading 4 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 65536 times 16384 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 989472 usec @ 1085.166 MiB/sec
Compare results ...

### (3.8) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4318134 usec @ 248.659 MiB/sec
# Reading 8 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 32768 times 32768 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 960031 usec @ 1118.445 MiB/sec
Compare results ...

### (3.16) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4318747 usec @ 248.623 MiB/sec
# Reading 16 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 16384 times 65536 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 957600 usec @ 1121.284 MiB/sec
Compare results ...

### (3.32) Writing 2 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Writing NVMe from cblk_read2.bin
Writing of 1073741824 bytes took 4331422 usec @ 247.896 MiB/sec
# Reading 32 blocks ...
NVMe device has 262144 blocks of each 4096 bytes; 1024 MiB @ 8 threads
Reading 8192 times 131072 bytes: 1048576 KiB NVMe into cblk_read3.bin
Reading of 1073741824 bytes took 957131 usec @ 1121.834 MiB/sec
Compare results ...

SUCCESS
```
